File: engines/go/benchmark.go
package main

import (
	"encoding/base64"
	"fmt"
	"io/ioutil"
	"os"
	"regexp"
	"strconv"
	"time"
)

func decodeBase64(base64String string) (string, error) {
	decoded, err := base64.StdEncoding.DecodeString(base64String)
	if err != nil {
		return "", err
	}
	return string(decoded), nil
}

func readFile(filename string) (string, error) {
	data, err := ioutil.ReadFile(filename)
	if err != nil {
		return "", err
	}
	return string(data), nil
}

func measure(data string, pattern string, fullMatch bool) {
	start := time.Now()

	r, err := regexp.Compile(pattern)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: Failed to compile regex: %v\n", err)
		os.Exit(1)
	}

	var count int

	if fullMatch {
		// Full match: entire text must match the regex
		match := r.FindString(data)
		if match == data {
			count = 1
		} else {
			count = 0
		}
	} else {
		// Partial match: find all matches in the text
		matches := r.FindAllString(data, -1)
		count = len(matches)
	}

	elapsed := time.Since(start)
	elapsedMs := float64(elapsed) / float64(time.Millisecond)

	fmt.Printf("%.6f - %d\n", elapsedMs, count)
}

func main() {
	if len(os.Args) != 4 {
		fmt.Printf("Usage: %s <base64_regex> <filename> <match_mode>\n", os.Args[0])
		fmt.Println("  base64_regex: Base64-encoded regular expression")
		fmt.Println("  filename: Path to the file containing text to match")
		fmt.Println("  match_mode: 1 for full match, 0 for partial match")
		os.Exit(1)
	}

	// Decode the base64 regex
	regex, err := decodeBase64(os.Args[1])
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: Failed to decode base64 regex: %v\n", err)
		os.Exit(1)
	}

	// Read file content
	data, err := readFile(os.Args[2])
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: Cannot open file %s: %v\n", os.Args[2], err)
		os.Exit(1)
	}

	// Parse match mode
	matchMode, err := strconv.Atoi(os.Args[3])
	if err != nil || (matchMode != 0 && matchMode != 1) {
		fmt.Fprintf(os.Stderr, "Error: match_mode must be 0 or 1\n")
		os.Exit(1)
	}

	// Measure and output results
	measure(data, regex, matchMode == 1)
}

File: engines/cpp/benchmark.cpp
#define _POSIX_C_SOURCE 199309L

#include <chrono>
#include <fstream>
#include <iostream>
#include <iomanip>
#include <regex>
#include <string>
#include <vector>
#include <stdexcept>
#include <cstdlib>
#include <boost/regex.hpp>
#include <openssl/bio.h>
#include <openssl/evp.h>
#include <openssl/buffer.h>

std::string base64_decode(const std::string& input) {
    BIO *bio, *b64;
    int decode_length = input.length();
    std::vector<char> buffer(decode_length + 1);
    
    bio = BIO_new_mem_buf(input.c_str(), -1);
    b64 = BIO_new(BIO_f_base64());
    bio = BIO_push(b64, bio);
    
    BIO_set_flags(bio, BIO_FLAGS_BASE64_NO_NL);
    int output_length = BIO_read(bio, buffer.data(), decode_length);
    buffer[output_length] = '\0';
    
    BIO_free_all(bio);
    
    return std::string(buffer.data(), output_length);
}

std::string read_file(const std::string& filename) {
    std::ifstream file(filename);
    if (!file) {
        std::cerr << "Error: Cannot open file " << filename << std::endl;
        std::exit(1);
    }
    
    return std::string(std::istreambuf_iterator<char>(file), 
                      std::istreambuf_iterator<char>());
}

void measure(const std::string& data, const std::string& pattern, int full_match) {
    using clock = std::chrono::high_resolution_clock;
    const auto start = clock::now();
    
    unsigned count = 0;
    
    try {
        if (full_match) {
            // Full match: entire text must match the regex
            const REGEX_NAMESPACE::regex re(pattern);
            if (REGEX_NAMESPACE::regex_match(data, re)) {
                count = 1;
            }
        } else {
            // Partial match: find all matches in the text
            const REGEX_NAMESPACE::regex re(pattern);
            for (REGEX_NAMESPACE::sregex_token_iterator it(data.cbegin(), data.cend(), re), end{}; 
                 it != end; ++it) {
                count++;
            }
        }
    } catch (const std::exception& e) {
        std::cerr << "Error: Failed to process regex: " << e.what() << std::endl;
        std::exit(1);
    }
    
    const auto end = clock::now();
    const double elapsed = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count() * 1e-6;
    std::cout << std::fixed << std::setprecision(6) << elapsed << " - " << count << std::endl;
}

int main(int argc, char** argv) {
    if (argc != 4) {
        std::cout << "Usage: " << argv[0] << " <base64_regex> <filename> <match_mode>" << std::endl;
        std::cout << "  base64_regex: Base64-encoded regular expression" << std::endl;
        std::cout << "  filename: Path to the file containing text to match" << std::endl;
        std::cout << "  match_mode: 1 for full match, 0 for partial match" << std::endl;
        return 1;
    }
    
    // Decode the base64 regex
    std::string regex;
    try {
        regex = base64_decode(argv[1]);
    } catch (const std::exception& e) {
        std::cerr << "Error: Failed to decode base64 regex: " << e.what() << std::endl;
        return 1;
    }
    
    // Read file content
    std::string data = read_file(argv[2]);
    
    // Parse match mode
    int match_mode = std::atoi(argv[3]);
    if (match_mode != 0 && match_mode != 1) {
        std::cerr << "Error: match_mode must be 0 or 1" << std::endl;
        return 1;
    }
    
    // Measure and output results
    measure(data, regex, match_mode);
    
    return 0;
}
